### PR TITLE
feat(core): implement broker and cluster level quota

### PR DESCRIPTION
**Pull request by ShadowySpirits**

### CODE REVIEW

**Comment by superhx** [comment]
Add AutoMQ inject start & code for changed code or use extends or use another method to avoid merge conflict

File: core/src/main/scala/kafka/server/DynamicConfig.scala [hunk 1]
```diff
@@ -41,6 +41,15 @@ object DynamicConfig {
...
+      .define(QuotaConfigs.BROKER_QUOTA_PRODUCE_BYTES_CONFIG, DOUBLE, Double.MaxValue, MEDIUM, QuotaConfigs.BROKER_QUOTA_PRODUCE_BYTES_DOC)
+      .define(QuotaConfigs.BROKER_QUOTA_FETCH_BYTES_CONFIG, DOUBLE, Double.MaxValue, MEDIUM, QuotaConfigs.BROKER_QUOTA_FETCH_BYTES_DOC)
+      .define(QuotaConfigs.BROKER_QUOTA_REQUEST_PERCENTAGE_CONFIG, DOUBLE, Double.MaxValue, MEDIUM, QuotaConfigs.BROKER_QUOTA_REQUEST_PERCENTAGE_DOC)
+      .define(QuotaConfigs.BROKER_QUOTA_WHITE_LIST_USER_CONFIG, STRING, "", MEDIUM, QuotaConfigs.BROKER_QUOTA_WHITE_LIST_USER_DOC)
+      .define(QuotaConfigs.BROKER_QUOTA_WHITE_LIST_CLIENT_ID_CONFIG, STRING, "", MEDIUM, QuotaConfigs.BROKER_QUOTA_WHITE_LIST_CLIENT_ID_DOC)
+      .define(QuotaConfigs.BROKER_QUOTA_WHITE_LIST_LISTENER_CONFIG, STRING, "", MEDIUM, QuotaConfigs.BROKER_QUOTA_WHITE_LIST_LISTENER_DOC)
+      .define(QuotaConfigs.CLUSTER_QUOTA_TOPIC_COUNT_CONFIG, LONG, Long.MaxValue, MEDIUM, QuotaConfigs.CLUSTER_QUOTA_TOPIC_COUNT_DOC)
+      .define(QuotaConfigs.CLUSTER_QUOTA_PARTITION_COUNT_CONFIG, LONG, Long.MaxValue, MEDIUM, QuotaConfigs.CLUSTER_QUOTA_PARTITION_COUNT_DOC)
```
Comment (superhx) [inline core/src/main/scala/kafka/server/DynamicConfig.scala:52]: 可以参考 DynamicBrokerConfig 里面 AutoBalancer 的用法添加动态配置

File: metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java [hunk 2]
```diff
@@ -683,14 +683,33 @@ ControllerResult<CreateTopicsResponseData> createTopics(
...
+
+            topicCount++;
+            if (topicCount > topicCountQuota) {
+                topicErrors.put(topic.name(),
+                    new ApiError(Errors.THROTTLING_QUOTA_EXCEEDED, "Topic count quota exceeded: current quota is " + topicCountQuota));
+            }
+        }
+
+        int partitionCountQuota = configurationControl.partitionCountQuota();
+        int partitionCount = this.topics.values().stream().mapToInt(info -> info.parts.size()).sum();
```
Comment (superhx) [inline metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:704]: Add inject end and check others

File: metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java [hunk 3]
```diff
@@ -716,18 +735,21 @@ ControllerResult<CreateTopicsResponseData> createTopics(
             } else {
                 configRecords = configResult.records();
             }
-            // AutoMQ for Kafka inject end
 
             ApiError error;
             try {
-                error = createTopic(context, topic, records, successes, configRecords, describable.contains(topic.name()));
+                error = createTopic(context, topic, records, successes, configRecords, describable.contains(topic.name()), partitionCountQuota, partitionCount);
```
Comment (superhx) [inline metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:741]: We could check the quota before #createTopic

**Review by superhx — DISMISSED** [review]

File: metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java [hunk 4]
```diff
@@ -683,6 +683,31 @@ ControllerResult<CreateTopicsResponseData> createTopics(
         Map<ConfigResource, Map<String, Entry<OpType, String>>> configChanges =
             computeConfigChanges(topicErrors, request.topics());
 
+        // AutoMQ inject start
+        // Verify cluster quota
+        int topicCountQuota = configurationControl.topicCountQuota();
```
Comment (daniel-y) [inline metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:688]: Extract this logic to an individual method?

Comment (daniel-y) [reply metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:688]: How about like this:

```java
private void checkQuota() {
        // .....
    }

    ControllerResult<CreateTopicsResponseData> createTopicsWithQuotaChecker(
        ControllerRequestContext context,
        CreateTopicsRequestData request,
        Set<String> describable
    ) {
        checkQuota();
        return createTopics(context, request, describable);
    }
```

**Review by superhx — APPROVED** [review]

**Pull request merged**